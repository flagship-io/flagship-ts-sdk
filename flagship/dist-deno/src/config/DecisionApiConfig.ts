import { DecisionMode, FlagshipConfig, IFlagshipConfig } from './FlagshipConfig.ts'

export class DecisionApiConfig extends FlagshipConfig {
  public constructor (param?: Omit<IFlagshipConfig, 'decisionMode'>) {
    super({ ...param, decisionMode: DecisionMode.DECISION_API })
  }
}
