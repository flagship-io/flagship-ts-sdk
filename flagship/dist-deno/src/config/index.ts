export * from './ConfigManager.ts'
export * from './DecisionApiConfig.ts'
export { BucketingConfig } from './BucketingConfig.ts'
export { DecisionMode, FlagshipConfig } from './FlagshipConfig.ts'
export type { IFlagshipConfig } from './FlagshipConfig.ts'
