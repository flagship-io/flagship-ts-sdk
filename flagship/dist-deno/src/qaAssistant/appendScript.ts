import { FS_QA_ASSISTANT_SCRIPT_TAG_ID, FS_QA_URL, TAG_QA_URL } from '../enum/FlagshipConstant.ts'

const getScriptElement = (src: string): HTMLScriptElement | null =>
  document.querySelector<HTMLScriptElement>(`script[src*="${src}"]`)
/**
 *
 * @param bundleFileUrl
 */
export function appendScript (bundleFileUrl: string): void {
  if (getScriptElement(TAG_QA_URL) || getScriptElement(FS_QA_URL)) {
    return
  }
  const script = document.createElement('script')
  script.src = bundleFileUrl
  script.id = FS_QA_ASSISTANT_SCRIPT_TAG_ID
  document.body.append(script)
}
