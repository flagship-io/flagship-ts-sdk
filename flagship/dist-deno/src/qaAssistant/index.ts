import { FS_QA_ASSISTANT, FS_QA_ASSISTANT_LOCAL, FS_QA_ASSISTANT_STAGING, QA_ASSISTANT_LOCAL_URL, QA_ASSISTANT_PROD_URL, QA_ASSISTANT_STAGING_URL, TAG_QA_ASSISTANT, TAG_QA_ASSISTANT_LOCAL, TAG_QA_ASSISTANT_STAGING } from '../enum/FlagshipConstant.ts'
import { IFlagshipConfig } from '../config/IFlagshipConfig.ts'
import { isBrowser } from '../utils/utils.ts'
import { listenForKeyboardQaAssistant } from './listenForKeyboardQaAssistant.ts'
import { loadQaAssistant } from './loadQaAssistant.ts'

/**
 *
 * @param config
 * @returns
 */
export function launchQaAssistant (config: IFlagshipConfig): void {
  if (!isBrowser()) {
    return
  }
  const urlMap = {
    [FS_QA_ASSISTANT]: QA_ASSISTANT_PROD_URL,
    [TAG_QA_ASSISTANT]: QA_ASSISTANT_PROD_URL,
    [FS_QA_ASSISTANT_STAGING]: QA_ASSISTANT_STAGING_URL,
    [TAG_QA_ASSISTANT_STAGING]: QA_ASSISTANT_STAGING_URL,
    [FS_QA_ASSISTANT_LOCAL]: QA_ASSISTANT_LOCAL_URL,
    [TAG_QA_ASSISTANT_LOCAL]: QA_ASSISTANT_LOCAL_URL
  }
  const queryParam = new URLSearchParams(window.location.search)
  const urlKey = Object.keys(urlMap).find(key => queryParam.get(key) === 'true') || ''

  if (config.isQAModeEnabled || urlKey) {
    loadQaAssistant(config, urlMap[urlKey as keyof typeof urlMap])
    return
  }
  listenForKeyboardQaAssistant(config)
}
