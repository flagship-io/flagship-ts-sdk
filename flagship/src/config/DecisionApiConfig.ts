import {FlagshipConfig} from "./FlagshipConfig.ts";
import {DecisionMode} from "../enum/DecisionMode.ts";

export class DecisionApiConfig extends FlagshipConfig{
    public constructor(envId?: string, apiKey?: string) {
        super(envId, apiKey);
        this.decisionMode = DecisionMode.DECISION_API
    }
}