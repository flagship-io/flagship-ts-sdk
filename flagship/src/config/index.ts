export * from './ConfigManager';
export * from './DecisionApiConfig';
export { BucketingConfig } from './BucketingConfig';
export { FlagshipConfig } from './FlagshipConfig';
export { DecisionMode } from './DecisionMode';
export type { IEdgeConfig } from './IEdgeConfig';
export type { IFlagshipConfig } from './IFlagshipConfig';
export type { ITrackingManagerConfig } from './TrackingManagerConfig';
