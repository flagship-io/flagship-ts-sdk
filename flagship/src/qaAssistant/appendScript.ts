import { FS_QA_ASSISTANT_SCRIPT_TAG_ID } from '../enum/FlagshipConstant'

/**
 *
 * @param bundleFileUrl
 */
export function appendScript (bundleFileUrl: string): void {
  const script = document.createElement('script')
  script.src = bundleFileUrl
  script.id = FS_QA_ASSISTANT_SCRIPT_TAG_ID
  document.body.append(script)
}
