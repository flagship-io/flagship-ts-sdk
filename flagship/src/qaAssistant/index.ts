import { FS_QA_ASSISTANT, FS_QA_ASSISTANT_LOCAL, FS_QA_ASSISTANT_STAGING, QA_ASSISTANT_LOCAL_URL, QA_ASSISTANT_PROD_URL, QA_ASSISTANT_STAGING_URL, TAG_QA_ASSISTANT, TAG_QA_ASSISTANT_LOCAL, TAG_QA_ASSISTANT_STAGING } from '../enum/FlagshipConstant'
import { IFlagshipConfig } from '../config/IFlagshipConfig'
import { isBrowser, onDomReady } from '../utils/utils'
import { listenForKeyboardQaAssistant } from './listenForKeyboardQaAssistant'
import { loadQaAssistant } from './loadQaAssistant'

/**
 *
 * @param config
 * @returns
 */
export function launchQaAssistant (config: IFlagshipConfig): void {
  if (!isBrowser()) {
    return
  }

  onDomReady(() => {
    const urlMap = {
      [FS_QA_ASSISTANT]: QA_ASSISTANT_PROD_URL,
      [TAG_QA_ASSISTANT]: QA_ASSISTANT_PROD_URL,
      [FS_QA_ASSISTANT_STAGING]: QA_ASSISTANT_STAGING_URL,
      [TAG_QA_ASSISTANT_STAGING]: QA_ASSISTANT_STAGING_URL,
      [FS_QA_ASSISTANT_LOCAL]: QA_ASSISTANT_LOCAL_URL,
      [TAG_QA_ASSISTANT_LOCAL]: QA_ASSISTANT_LOCAL_URL
    }
    const queryParam = new URLSearchParams(window.location.search)
    const urlKey = Object.keys(urlMap).find(key => queryParam.get(key) === 'true') || ''

    if (config.isQAModeEnabled || urlKey) {
      loadQaAssistant(config, urlMap[urlKey as keyof typeof urlMap])
      return
    }
    listenForKeyboardQaAssistant(config)
  })
}
