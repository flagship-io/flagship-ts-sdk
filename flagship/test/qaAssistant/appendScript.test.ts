/**
 * @jest-environment jsdom
 */

import { describe, it, expect } from '@jest/globals';
import { DecisionApiConfig } from '../../src/config/DecisionApiConfig';
import { appendScript } from '../../src/qaAssistant/appendScript';
import { FS_QA_ASSISTANT_SCRIPT_TAG_ID,
  QA_ASSISTANT_LOCAL_URL,
  QA_ASSISTANT_PROD_URL } from '../../src/enum/FlagshipConstant';

describe('Test appendScript', () => {
  const config = new DecisionApiConfig();
  config.envId = 'envId';

  it('should append production script to the document', () => {
    appendScript(QA_ASSISTANT_PROD_URL);

    const tagScript = document.getElementById(
      FS_QA_ASSISTANT_SCRIPT_TAG_ID
    ) as HTMLScriptElement | null;
    expect(tagScript).toBeDefined();
    expect(tagScript?.src).toBe(QA_ASSISTANT_PROD_URL);
    expect(tagScript?.crossOrigin).toBe('');
    expect(tagScript).toMatchSnapshot();
  });

  it('should append local script to the document with anonymous cross-origin', () => {
    let tagScript = document.getElementById(
      FS_QA_ASSISTANT_SCRIPT_TAG_ID
    ) as HTMLScriptElement | null;
    tagScript?.remove();

    appendScript(QA_ASSISTANT_LOCAL_URL);

    tagScript = document.getElementById(
      FS_QA_ASSISTANT_SCRIPT_TAG_ID
    ) as HTMLScriptElement | null;
    expect(tagScript).toBeDefined();
    expect(tagScript?.src).toBe(QA_ASSISTANT_LOCAL_URL);
    expect(tagScript?.crossOrigin).toBe('anonymous');
    expect(tagScript).toMatchSnapshot();
  });
});
