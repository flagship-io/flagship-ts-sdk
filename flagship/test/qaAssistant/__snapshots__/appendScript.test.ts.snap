// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test appendScript test Script is appended to document 1`] = `
<script
  id="FS_QA_ASSISTANT_SCRIPT_TAG_ID"
  src="https://qa-assistant.abtasty.com/bundle.js"
/>
`;
